hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/common/AbstractCaseOperator.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate.operators.common;
+
+import com.hazelcast.sql.impl.calcite.validate.HazelcastCallBinding;
+import com.hazelcast.sql.impl.calcite.validate.operators.HazelcastCaseOperator.CaseReturnTypeInference;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlCaseOperator;
+import org.apache.calcite.sql.type.InferTypes;
+
+import static com.hazelcast.sql.impl.calcite.validate.operators.HazelcastReturnTypeInference.wrap;
+
+public abstract class AbstractCaseOperator extends SqlOperator implements HazelcastOperandTypeCheckerAware {
+
+    protected AbstractCaseOperator() {
+        super(SqlCaseOperator.INSTANCE.getName(), SqlKind.CASE, SqlCaseOperator.INSTANCE.getLeftPrec(), true, wrap(new CaseReturnTypeInference()),
+                InferTypes.RETURN_TYPE, null);
```

Comment:
I doubt we should use `InferTypes.RETURN_TYPE` because it doesn't seem to do the right thing. In the `CASE` operator, we have two types of operands: conditions and return branches. What we should do is:
1. Assign the `BOOLEAN` type to all conditions
2. infer the types of all return branches. If some branch has an UNKNOWN type (e.g. because it is an argument), then assign the type with the highest precedence across all return branches with known types
3. If all return branches have the `UNKNOWN` type, then throw a validation error.

Example 1:
`CASE WHEN x>1 THEN int_col ELSE ?` - should return [BOOLEAN, INTEGER, INTEGER]

Example 2:
`CASE WHEN ? THEN int_col1 ELSE int_col2` - should return [BOOLEAN, INTEGER, INTEGER]

Example 3:
`CASE WHEN x>? THEN ? ELSE ?` - should throw an error, because all return branches have the `UNKNOWN` type.

See `ReplaceUnknownOperandTypeInference` for reference.